### PR TITLE
68 multiple entries on go to definition

### LINKED ISSUE #68: Multiple Entries on Go to Definition
In the [`app/page.tsx` file](https://github.com/Viijay-Kr/react-ts-css/blob/main/examples/next-js-app-dir-css-modules/app/page.tsx) in [the Next.js app dir example](https://github.com/Viijay-Kr/react-ts-css/tree/main/examples/next-js-app-dir-css-modules) `cmd`-clicking on the `styles.code` property to trigger Go to Definition in VS Code shows an extra entry of the [`node_modules/next/types/global.d.ts`](https://github.com/vercel/next.js/blob/77f40ae9f1510ebb306d6c1563b989359b10190b/packages/next/types/global.d.ts#L21-L34) file with the `*.module.css` type definition.

Screenshot 2023-01-26 at 12 22 52

This makes it no longer a one-click Go to Definition interaction. It would be great to get rid of these generic `*.module.css` / `*.module.scss` / `*.module.sass` entries so that it's a single click again.

This is also a problem with other extensions:

- https://github.com/clinyong/vscode-css-modules/issues/63

### Potential approach

It seems that @zardoy has found a potentially interesting approach in his extension [TypeScript Essential Plugins](https://marketplace.visualstudio.com/items?itemName=zardoy.ts-essential-plugins) ([repo](https://github.com/zardoy/typescript-vscode-plugins)). Some interesting parts of the code:

```ts
    proxy.getDefinitionAndBoundSpan = (fileName, position) => {
      const prior = info.languageService.getDefinitionAndBoundSpan(fileName, position)
      // ...
      prior.definitions = prior.definitions.filter(({ fileName, containerName, containerKind, kind, name, ...rest }) => {
        // ...
        if (moduleDeclaration?.name.text === '*.module.css') return false
        return true
      })
      // ...
    })
```

- https://github.com/zardoy/typescript-vscode-plugins/blob/a102a76786a0292aa8a9a46fb6cc43595e8871a7/typescript/src/definitions.ts#L7-L8
- https://github.com/zardoy/typescript-vscode-plugins/blob/a102a76786a0292aa8a9a46fb6cc43595e8871a7/typescript/src/definitions.ts#L123
- https://github.com/zardoy/typescript-vscode-plugins/blob/a102a76786a0292aa8a9a46fb6cc43595e8871a7/typescript/src/definitions.ts#L141

If I'm understanding correctly, this is filtering out previous definitions if they have the name `*.module.css`, which could also be extended to include `*.module.scss` and `*.module.sass`.

### PR DESCRIPTION
**Pull request by Viijay-Kr**
Closes #68 

@karlhorky Here is the first draft to help avoid the extra definitions results from declaration files

I m using a small [typescript plugin](https://www.npmjs.com/package/typescript-cleanup-definitions) to help this cause.

currently the option to override the definitions from declaration files can be done to settings Which is documented in the readme

Let me know what you think about this PR

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -30,6 +33,16 @@
...
       "preLaunchTask": "tasks: watch-tests"
+    },
+    {
+      "name": "Attach to TS Server",
+      "type": "node",
+      "request": "attach",
+      "port": 5667,
+      "sourceMaps": true,
+      "outFiles": [
+        "/Volumes/VijayK/private/rtcss-tsplugin/node_modules/typescript/bin/tsserver"
```
Comment (karlhorky) [inline .vscode/launch.json:44]: Should this path be hard coded?

File: README.md [hunk 2]
```diff
@@ -15,6 +15,12 @@ This extension also supports CSS language features which are not supported by bu
 > - Nested selectors
 > - Suffixed selectors ([SCSS only](https://sass-lang.com/documentation/style-rules/parent-selector#adding-suffixes))
 > - Deeply nested suffix selectors
+>
+> Almost all the project scaffolders such as vite , next js and CRA adds css module declration  to the project by injecting it in '.d.ts' file (for instance inside `node_modules/vite/client.d.ts` added by vite,). Typescript treats these definitions as definition provider for Style properties. This results in a useless defintion result when VS `Code Go to Definition` is tiggered. Check this [issue](https://github.com/Viijay-Kr/react-ts-css/issues/68).
```
Comment (karlhorky) [inline README.md:19]:
```suggestion
> Almost all project scaffolders such as Vite, Next.js and CRA add css module declaration to the project by injecting it in a `.d.ts` file (for instance inside `node_modules/vite/client.d.ts` added by Vite). TypeScript treats these definitions as definition provider for Style properties. This results in a useless definition result when VS Code `Go to Definition` is triggered. Check this [issue](https://github.com/Viijay-Kr/react-ts-css/issues/68).
```

File: README.md [hunk 3]
```diff
@@ -15,6 +15,12 @@ This extension also supports CSS language features which are not supported by bu
...
+> 
+> This extension gives you an option to avoid that  result  using a typescript plugin [typescript-cleanup-defs](https://www.npmjs.com/package/typescript-cleanup-definitions) that can filter out those definitions results. Check the plugin for more details
```
Comment (karlhorky) [inline README.md:21]:
```suggestion
> This extension gives you an option to eliminate the useless results by using the TypeScript plugin [typescript-cleanup-defs](https://www.npmjs.com/package/typescript-cleanup-definitions) that can filter out those definitions results. Check the plugin for more details.
```

File: examples/next-js-app-dir-css-modules/app/page.module.css [hunk 4]
```diff
@@ -253,7 +253,7 @@
 @media (prefers-color-scheme: dark) {
   .vercelLogo {
     filter: invert(1);
-  }
+  },
```
Comment (karlhorky) [inline examples/next-js-app-dir-css-modules/app/page.module.css:256]: Is this an error?

File: package.json [hunk 5]
```diff
@@ -91,26 +91,49 @@
...
+          "description": "Typescript cleanup definitions plugin can be disabled with this setting. 'reactTsScss.cleanUpDefs' will have no effect when this setting is false",
+          "default": true
+        },
+        "reactTsScss.cleanUpDefs": {
+          "type": "array",
+          "title": "Typescript Cleanup Definitions",
+          "scope": "window",
+          "description": "Add a declaration file to skip definition results from this file. Useful for skipping unnecessary defintions from boilder plate projects",
+          "default": [
+            "node_modules/vite/client.d.ts",
```
Comment (karlhorky) [inline package.json:124]: Will these paths work at any level of the project?

Eg. In a monorepo / workspace project, the path would be:

```
<project root>/packages/package-name/node_modules/next/types/global.d.ts
```

Would be cool if these also worked out of the box without any additional configuration

**Review by karlhorky — CHANGES_REQUESTED** [review]
Seems like a really cool approach!

I suggested a few changes and asked some questions

File: .vscode/launch.json (hunk 1 above)
Comment (Viijay-Kr) [reply .vscode/launch.json:44]: This got in by mistake

File: package.json (hunk 5 above)
Comment (Viijay-Kr) [reply package.json:124]: it should since the plugin does a vague/naive string comparison .

**Comment by Viijay-Kr** [comment]
> Seems like a really cool approach!
> 
> I suggested a few changes and asked some questions

Thanks a lot for the comments @karlhorky . Really Appreciate your time and effort.

File: examples/next-js-app-dir-css-modules/app/page.module.css (hunk 4 above)
Comment (Viijay-Kr) [reply examples/next-js-app-dir-css-modules/app/page.module.css:256]: yeah good catch

File: package.json (hunk 5 above)
Comment (Viijay-Kr) [reply package.json:124]: @karlhorky Added a mono repo setup in the latest commit to test this out . Tested both in vite and next app and tt works as I expected.

you can also test it . 
Open the monorepo folder in debug mode. Open `App.tsx` inside `examples/monorepo/packages/vite-app/src/App.tsx` and try going to the definition of `styles.logo`. You'll see only the css module definition is provided :)

**Comment by karlhorky** [comment]
I tried this out just now in `v1.6.2`, but I'm still getting the duplicate entries:

Screenshot 2023-02-01 at 19 05 45


More details:

- Page: `packages/fusion.upleveled.io/app/(auth)/login/page.tsx`
- CSS Module File: `packages/fusion.upleveled.io/app/(auth)/login/page.module.scss`
- Next.js type definitions: `node_modules/next/types/global.d.ts`

**Comment by karlhorky** [comment]
It seems like in [the `monorepo` example](https://github.com/Viijay-Kr/react-ts-css/tree/main/examples/monorepo), the `packages/next-app` workspace is not configured in `package.json`, wonder if this is related:

https://github.com/Viijay-Kr/react-ts-css/blob/19198f9c574550011a1df3e61ed2df3e04c91cfb/examples/monorepo/package.json#L5-L10

**Comment by karlhorky** [comment]
If I should open a new issue, let me know.

**Comment by Viijay-Kr** [comment]
@karlhorky I too tried it after the version was shipped . It did not work for me either. I suspect it could be something to do with the typescript version difference .

I did not have time to investigate this further.

Hope I will find some time to debug this.

**Comment by karlhorky** [comment]
I'll create an issue so that we don't lose track of it.

**Comment by Viijay-Kr** [comment]
> It seems like in [the `monorepo` example](https://github.com/Viijay-Kr/react-ts-css/tree/main/examples/monorepo), the `packages/next-app` workspace is not configured in `package.json`, wonder if this is related:
> 
> https://github.com/Viijay-Kr/react-ts-css/blob/19198f9c574550011a1df3e61ed2df3e04c91cfb/examples/monorepo/package.json#L5-L10
> 
> Eg. could that mean that the `node_modules/next/types/global.d.ts` is not being hoisted to the root?

This is not related . 

I have removed the next app from packages in order for the package to have it own node_modules folder . 

If you look at vite example it has its own node modules setup.So this helped to test the use case you mentioned

However removing the next app from the list of packages didn't make a difference since the monorepo was setup using turbo repo which provides a next boilerplate

**Comment by karlhorky** [comment]
I reopened #68 and added a comment: https://github.com/Viijay-Kr/react-ts-css/issues/68#issuecomment-1412524350

**Pull request merged**